### PR TITLE
Replace size check with more meaningful pubkey check

### LINKED ISSUE #15465: OpenSSL 3.0.0alpha16 interop issue with 0.9.8zh - Diffie-Hellman invalid public key
I ran into an interop problem when doing release testing for httpd 2.4.48 using OpenSSL 3.0.0alpha16 in the server and OpenSSL 0.9.8zh in the client. First I made TLS 1.0 work by adding "@SECLEVEL=0" at the end of the SSLCipherSuite config lines, but I still see sporadic failures:

SSL Library Error: error:02800066:Diffie-Hellman routines::invalid public key

I think they might have to do with restricting the allowed params for DH in OpenSSL 3. So it seems sometimes 0.9.8 generates DH keys that do not fulfil the stricter requirements of version 3. That means using such old client software will most of the time succeed with an TLS 1.0 handshake, but sporadically it will fail. I am not arware of a workaround.

Of course TLS 1.0 should no longer be used, but since it is supported in OpenSSL 3, the behavior of sporadic interop failures is a bit problematic and should at least be documented.

With a bit of help how to add more debug info, I could do a custom httpd build. But I will try to reproduce with s_server and s_client first and hope there's enough debug flags there to understand the used DH params.

### PR DESCRIPTION
**Pull request by t8m**
It does not make sense to check the size because this
function can be used in other contexts than in TLS-1.3 and
the value might not be padded to the size of p.

However it makes sense to do the partial pubkey check because
there is no valid reason having the pubkey value outside the
1 < pubkey < p-1 bounds.

Fixes #15465

### CODE REVIEW

File: crypto/dh/dh_key.c [hunk 1]
```diff
@@ -375,20 +375,17 @@ int ossl_dh_buf2key(DH *dh, const unsigned char *buf, size_t len)
...
     DH_get0_pqg(dh, &p, NULL, NULL);
-    if (p == NULL || (p_size = BN_num_bytes(p)) == 0) {
+    if (p == NULL || BN_num_bytes(p) == 0) {
         err_reason = DH_R_NO_PARAMETERS_SET;
         goto err;
     }
-    /*
-     * As per Section 4.2.8.1 of RFC 8446 fail if DHE's
-     * public key is of size not equal to size of p
-     */
```
Comment (mattcaswell) [inline crypto/dh/dh_key.c:390]: The referenced TLS RFC still seems relevant but perhaps not here? Should we be checking whether p is padded correctly in libssl instead? The RFC doesn't actually say that peers should validate this, but it does say the endpoints emitting these keys should send them padded.

Comment (t8m) [reply crypto/dh/dh_key.c:390]: I would not bother validating this. I cannot see any issue coming from this missing size validation.

**Comment by t8m** [comment]
I am working on testcases for the EVP_PKEY_set1_encoded_public_key()

**Review by paulidale — APPROVED** [review]

File: crypto/dh/dh_key.c (hunk 1 above)
Comment (kaduk) [reply crypto/dh/dh_key.c:390]: Checking that the key share is padded to the proper encoded length is part of being a good citizen in the broader ecosystem.  If enough implementations are lax in checking, then implementations that are lax in sending can escape into the wild and force a burden of compatibility (accepting the incorrectly padded values) onto all implementations.
But yes, this particular padding check is more appropriate in libssl than libcrypto.

**Pull request closed without merging**